internal/core/src/index/StringIndexSort.cpp

```diff
@@ -0,0 +1,1129 @@
+    auto total_string_size = 0;
+    const uint8_t* last_str_ptr =
+        result.string_data_start +
+        result.string_offsets[result.unique_count - 1];
+    uint32_t last_str_len;
+    memcpy(&last_str_len, last_str_ptr, sizeof(uint32_t));
+    total_string_size = result.string_offsets[result.unique_count - 1] +
+                        sizeof(uint32_t) + last_str_len;
+
+    // Skip past string section to posting list offsets
+    ptr = result.string_data_start + total_string_size;
+    result.post_list_offsets = reinterpret_cast<const uint32_t*>(ptr);
+    ptr += result.unique_count * sizeof(uint32_t);
+
+    result.post_list_data_start = ptr;
+
+    return result;
+}
+
+const std::string STRING_INDEX_SORT_FILE = "string_index_sort";
+
+StringIndexSort::StringIndexSort(
+    const storage::FileManagerContext& file_manager_context)
+    : StringIndex(ASCENDING_SORT), is_built_(false) {
+    if (file_manager_context.Valid()) {
+        field_id_ = file_manager_context.fieldDataMeta.field_id;
+        file_manager_ =
+            std::make_shared<storage::MemFileManagerImpl>(file_manager_context);
+    }
+}
+
+StringIndexSort::~StringIndexSort() = default;
+
+int64_t
+StringIndexSort::Count() {
+    return total_num_rows_;
+}
+
+void
+StringIndexSort::Build(size_t n,
+                       const std::string* values,
+                       const bool* valid_data) {
+    if (is_built_)
+        return;
+    if (n == 0) {
+        ThrowInfo(DataIsEmpty, "StringIndexSort cannot build null values!");
+    }
+
+    index_build_begin_ = std::chrono::system_clock::now();
+    total_num_rows_ = n;
+    valid_bitset_ = TargetBitmap(total_num_rows_, false);
+    idx_to_offsets_.clear();
+
+    // Create MemoryImpl and delegate building to it
+    impl_ = std::make_unique<StringIndexSortMemoryImpl>();
+    auto* memory_impl = static_cast<StringIndexSortMemoryImpl*>(impl_.get());
+
+    // Let MemoryImpl handle the building process
+    memory_impl->BuildFromRawData(
+        n, values, valid_data, valid_bitset_, idx_to_offsets_);
+
+    is_built_ = true;
+    total_size_ = CalculateTotalSize();
+}
+
+void
+StringIndexSort::Build(const Config& config) {
+    if (is_built_) {
+        return;
+    }
+    config_ = config;
+    auto field_datas =
+        storage::CacheRawDataAndFillMissing(file_manager_, config);
+    BuildWithFieldData(field_datas);
+}
+
+void
+StringIndexSort::BuildWithFieldData(
+    const std::vector<FieldDataPtr>& field_datas) {
+    if (is_built_)
+        return;
+
+    index_build_begin_ = std::chrono::system_clock::now();
+
+    // Calculate total number of rows
+    total_num_rows_ = 0;
+    for (const auto& data : field_datas) {
+        total_num_rows_ += data->get_num_rows();
+    }
+
+    if (total_num_rows_ == 0) {
+        ThrowInfo(DataIsEmpty, "StringIndexSort cannot build null values!");
+    }
+
+    // Initialize structures
+    valid_bitset_ = TargetBitmap(total_num_rows_, false);
```

Comment:
I see you have `idx_to_offsets_.clear();` in `Build(size_t n, const std::string* values, const bool* valid_data)` but not here